File: java/algorithms/sort/quickSort/QuickSort1.java
import java.util.Arrays;

public class QuickSort1 {
    public static void main(String[] args) {
        int[] arr = new int[] { 23, 13, 8, 30, 29, 9 };
        quickSort(arr, 0, arr.length - 1);
        Arrays.stream(arr).forEach(System.out::println);
    }

    private static void quickSort(int[] arr, int l, int r) {
        if (l >= r) {
            return;
        }

        int pi = partition(arr, l, r);
        quickSort(arr, l, pi - 1);
        quickSort(arr, pi + 1, r);
    }

    private static int partition(int[] arr, int l, int r) {
        int pivot = arr[r];
        int i = l - 1; // <-- why is this important to start here?
        for (int j = l; j < r; j++) {
            if (arr[j] <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, r);
        return i + 1;
    }

    private static void swap(int[] arr, int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }
}

File: python/algorithms/sort/practice/quicksort_1.py
import random


def quicksort(l, r, arr):
    if r <= l:
        return
    
    p = partition(l, r, arr)
    quicksort(l, p - 1, arr)
    quicksort(p + 1, r, arr)
    return arr

def partition(l, r, arr):
    p = random.randint(l, r)
    arr[p], arr[r] = arr[r], arr[p]
    i = l - 1
    for j in range(l, r):
        if arr[j] <= arr[r]:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[r] = arr[r], arr[i + 1]
    return i + 1

nums = [2, 5, 8, 1, 9, 4, 3]
print(quicksort(0, len(nums) - 1, nums))